source/h2_decoder.c

```diff
@@ -752,7 +774,25 @@ static int s_state_fn_frame_settings_i(struct aws_h2_decoder *decoder, struct aw
     /* An endpoint that receives a SETTINGS frame with any unknown or unsupported identifier MUST ignore that setting.
      * RFC-7540 6.5.2 */
     if (id >= AWS_H2_SETTINGS_BEGIN_RANGE && id < AWS_H2_SETTINGS_END_RANGE) {
-        DECODER_CALL_VTABLE_ARGS(decoder, on_settings_i, id, value);
+        /* check the value meets the settings bounds */
+        if (value < aws_h2_settings_bounds[id][0] || value > aws_h2_settings_bounds[id][1]) {
+            DECODER_LOGF(
+                ERROR, decoder, "A value of SETTING frame is invalid, id: %" PRIu16 ", value: %" PRIu32, id, value);
+            aws_array_list_clean_up(&decoder->settings_buffer_list);
+            if (id == AWS_H2_SETTINGS_INITIAL_WINDOW_SIZE)
+                return aws_raise_error(AWS_H2_ERR_FLOW_CONTROL_ERROR);
+            else
+                return aws_raise_error(AWS_H2_ERR_PROTOCOL_ERROR);
```

Comment:
two things
1) trivial/style: we always use curly-braces, even for 1-line if-statements
2) important/confusing: raise `AWS_ERROR_HTTP_PROTOCOL_ERROR`, not `AWS_H2_ERR_PROTOCOL_ERROR`. aws_raise_error() expects one of our library error codes, it's going to think `AWS_H2_ERR_PROTOCOL_ERROR`(0x1) is `AWS_ERROR_OOM`(0x1). We haven't written the code yet, but there will be a function that translates errors from `AWS_ERROR_HTTP_XYZ` into `AWS_H2_ERR_XYZ` so we can send a GOAWAY with the appropriate error-code (anything that doesn't map to an H2 error code will become `AWS_H2_ERR_INTERNAL_ERROR`. We haven't declared AWS_ERROR_HTTP_XYZ for all the H2 codes yet. I was declaring that FLOW_CONTROL one in the PR I'm working on now. Put a TODO on that line and whoever merges first can fix it